Fix floating point initialization issue in VRF residual function

rongpengzhang opened the pull request:
Addresses #5374 - Uninitialized variables in VRF temperature control.

Possibly a divide by zero happens in "Pipe_Num_St =
Pipe_Num_Nu/Pipe_Num_Re/Pipe_Num_Pr;", so make it more robust to handle
the case.

rongpengzhang committed:
Handle possible debug build error

Possibly a divide by zero happens in "Pipe_Num_St =
Pipe_Num_Nu/Pipe_Num_Re/Pipe_Num_Pr;", so make it more robust to handle
the case.

Myoldmopar commented:
OK, I'm going to have lunch, we'll see if CI gets to this before me and confirms whether this fixes it or not.

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -7910,26 +7925,33 @@ namespace HVACVariableRefrigerantFlow {
+				Pipe_Q = 0;
+			}
+
 			Pipe_h_comp_out = Pipe_h_IU_in + Pipe_Q / Pipe_m_ref / 3600; 
```
And if Pipe_m_ref = 0 here?

I was thinking more like this higher up in the function:

```
if ( Pipe_m_ref == 0 ) {
  initialize all necessary report and other variables
  return;
}
```

rongpengzhang replied:
@rraustad that's a more organized design. But first let's see whether the crash really comes from the "divide by 0".  We are still not sure of it. Quite possible this is not the reason, since it runs fine on my Windows machine. Maybe it is due to the parameter initialization. 

If "divide by 0" is really the reason, I will implement the code organization you suggested. If no time to do it now, I will definitely do that in the VRF-HR development, in which I will need to re-design the CalcVRFCondenser_FluidTCtrl method.

rongpengzhang commented:
@Myoldmopar  I just talked with @wfbuhl, and he mentioned that another possible reason can come from the parameter initialization. If it still fails, I will double check that. 

Anyway, l think we can fix it today.That should be quick. 

Enjoy your lunch!

DeadParrot commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -7910,26 +7925,33 @@ namespace HVACVariableRefrigerantFlow {
 			
-			Pipe_Q = max( 0.0, ( 3.141593 * VRF( VRFCond ).RefPipLen ) * ( Pipe_T_IU_in - OutdoorDryBulb / 2 - Pipe_T_room / 2 ) / ( 1 / Pipe_Coe_k1 + 1 / Pipe_Coe_k2 + 1 / Pipe_Coe_k3 ) ); // [W] 
-			Pipe_DeltP = max( 0.0, 8 * Pipe_Num_St * std::pow( Pipe_Num_Pr, 0.6667) * VRF( VRFCond ).RefPipEquLen / VRF( VRFCond ).RefPipDia * GetSupHeatDensityRefrig( VRF( VRFCond ).RefrigerantName, 
+				Pipe_Q = max( 0.0, ( 3.141593 * VRF( VRFCond ).RefPipLen ) * ( Pipe_T_IU_in - OutdoorDryBulb / 2 - Pipe_T_room / 2 ) / ( 1 / Pipe_Coe_k1 + 1 / Pipe_Coe_k2 + 1 / Pipe_Coe_k3 ) ); // [W] 
```
Please use `DataGlobals ::Pi` instead of the (single precision) literal here and elsewhere in this file.

rongpengzhang replied:
Thanks @DeadParrot. I'll keep this in mind. CalcVRFCondenser_FluidTCtrl was designed by my colleague, and in my opinion it is too long. I plan to re-organize it when I develop the VRF-HR codes in the following months. I will also separate the piping loss calculations as an individual method.

rongpengzhang commented:
@Myoldmopar It shows "All checks have passed", but in fact it is only tested on MacOS.

So far so good. No problem comes:
![image](https://cloud.githubusercontent.com/assets/7785946/11855167/6d6a6f12-a3fd-11e5-84fe-a93108a2c2e7.png)

Is there a faster way to test this commit on Linux?

rongpengzhang committed:
Handle possible debug build error

A bit more codes to handle the case Pipe_m_ref=0. I think this
does not happen at all at normal simulations. Maybe it happens during
initialization, and it is only considered as an error on Linux, not
Windows or Mac?

Myoldmopar commented:
> @Myoldmopar It shows "All checks have passed", but in fact it is only tested on MacOS.

Yes.  GitHub has _no_ idea how many tests are going to be run.  So far Marvin alerted GitHub that it was running one, and then he told GitHub he was done.  So GitHub says everything thus far is done.

> Is there a faster way to test this commit on Linux?

Yes, of course.  Build and test locally.  I can do it. I got pulled aside, but I'll kick off a build now.

mjwitte commented:
Has anyone else notice a lag in e-mail notification from github today?  @Myoldmopar post a comment at 4:26 CST and I got the e-mail at 4:50.

Myoldmopar commented:
Everything looks OK according to [github status](https://status.github.com/messages), and I haven't been watching close enough to notice.

rongpengzhang commented:
@mjwitte I noticed the lag too. Email is about 25min later.

Myoldmopar commented:
@rongpengzhang This fix does _not_ stop the hard crash.

rongpengzhang commented:
@Myoldmopar  Did you test with the codes in commit 8800ce5, which are newer than those in c0a1b63? 

If so, can you provide a bit more information on the crash? At which line it crashed? Due to "divide by zero" or something else?

rongpengzhang commented:
@Myoldmopar @mjwitte 

It seems it runs fine on Linux-Ubuntu now. 
https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-c0a1b6312259a022bad6c21e1d611ee150693553-x86_64-Linux-Ubuntu-14.04-gcc-4.8.html

Now "VariableRefrigerantFlow_FluidTCtrl_5Zone.idf" failed because of result diff, which is reasonable.
![image](https://cloud.githubusercontent.com/assets/7785946/11857223/3aa35f2c-a40b-11e5-9f2c-5ec92ac8c44e.png)

Before the fix, it failed when "initializing simulation"
https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-9ff9bbb09e88e79f373beb6a1e852aafe19c857d-x86_64-Linux-Ubuntu-14.04-gcc-4.8-IntegrationCoverage-Debug.html#test
![image](https://cloud.githubusercontent.com/assets/7785946/11857238/620cfc62-a40b-11e5-8d02-d2ba8628d1e2.png)

Myoldmopar commented:
Not quite.  The release build succeeds.  The integration-coverage-DEBUG will fail when CI gets to that platform/configuration.

I used commit 8800ce5.

Fails on line 9082: https://github.com/NREL/EnergyPlus/blob/8800ce58f91592e7b59eca1ff1a4e8f9ff9b6666/src/EnergyPlus/HVACVariableRefrigerantFlow.cc#L9082

Myoldmopar committed:
Fix Par index

Myoldmopar commented:
So, I am getting a floating point exception on line 9082 @ 8800ce5:

```
FirstHVACIteration = ( Par( 1 ) == 1.0 );
```

Where Par is a function argument of type:

```
Array1< Real64 > const & Par
```

For kicks I tried changing that line so it is:

```
if ( Par( 1 ) < 0.001 ) {
  FirstHVACIteration = false;
} else {
  FirstHVACIteration = true;
}
```

and it still gets an FPE when evaluating the `Par( 1 )` line.  

I thought for a moment maybe the argument type was wrong for that ObjexxFCL array instance.  But nope, it matches others that work.

So I debugged in, and found that Par( 1 ) is NaN.  WtF?  It's assigned right [here](https://github.com/NREL/EnergyPlus/blob/8800ce58f91592e7b59eca1ff1a4e8f9ff9b6666/src/EnergyPlus/HVACVariableRefrigerantFlow.cc#L8967-L8971) before it's called.  See on first hvac, `Par( 1 ) = 1;` and not on first hvac `Par( 2 ) = 0;`.......... wait what?  Yeah, there it is.

Rebuilding now, testing, yep, fixed.  Pushed as 4971f9a.

FYI @rongpengzhang @mjwitte @lgentile Once CI comes back clean here, this will be merged and this completes the E+ source changes for the bug fix release.  I have a couple bookkeeping things to fix, and that should be the final commit.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into #5252_TinDiscrepancyNewVRFModel

Myoldmopar commented:
So...why is this causing diffs where it looks like the field name changed being reported by the sizing routine in the eio?  @rongpengzhang

rongpengzhang commented:
@Myoldmopar 

> The issue only exists there because it's the only debug build, so it catches more.

But I always run VS in Debug mode on my local machine (Windows). It didn't report anything. Maybe I should use some more strict setting? Any ideas on that?

> So...why is this causing diffs where it looks like the field name changed being reported by the sizing routine in the eio?

The updated ones are correct. Rich pointed them out and I corrected them.

rongpengzhang commented:
@Myoldmopar Thanks for the careful debug! You have sharp eyes~

Myoldmopar commented:
@rongpengzhang The debug-build compiler flags that are set up by the CMake project as it is right now are apparently not as strict on Windows as they are on Linux.  So in Linux, the g++ flags will trap for more stuff.  I am looking at improving the testing by adding Windows debug builds, and adjusting flags.  If you make the flags more strict in Visual Studio, and you are able to reproduce the error found by the Ubuntu CI debug build, please send me the settings you changed so we can add them to the CMake file.  This would be most helpful.

OK on the field names.  That is good.  If the integration coverage debug build comes back clean, this should wrap up the bug-fix release as far as E+ source changes go.

rongpengzhang commented:
@Myoldmopar  

I've found a way to report FPE in VS:

1) In VS menu, Project -> Properties -> C/C++ -> Code Generation -> Floating Pint Model, change its setting to "Strict(/fp:strict)".
2) Use _controlfp to set the floating-point control word (see code below).

```
#include "stdafx.h"
#include <float.h>
unsigned int fp_control_state = _controlfp(_EM_INEXACT, _MCW_EM);

int _tmain(int argc, _TCHAR* argv[])
{
    double x = 0.0;
    double y = 1.0/x;   // floating point exception

    return 0;
}

```

Edwin, maybe you can try to reconfigure CMake to implement these settings when you have time.

FYI @mjwitte @wfbuhl

Myoldmopar commented:
So @rongpengzhang, if you just do the first item (fp:strict), are you then able to reproduce the error?  Maybe @mbadams5 or @DeadParrot could offer some guidance here?

mjwitte commented:
I haven't tested this recently, but I think the current Win debug settings will flag a divide by zero, but not a spurious NaN caused by use of an uninitialized variable.

Myoldmopar commented:
Good call on that @mjwitte; I had `div zero` in my head...

rongpengzhang commented:
@Myoldmopar I think both of them need to be implemented to report the error

rongpengzhang commented:
@mjwitte The following two cases are usually treated differently:
(1) int x = 0; int y = 1 / x;
(2) double x = 0.0; double y = 1 / x;

I think only (2) is considered as FPE, and cannot be reported. (1) can be reported even if I don't change any settings.

Does the current Win debug setting will flag for both of them?

mjwitte commented:
@rongpengzhang I'm not sure - I'm only going on memory from fixing bugs.  If you can make a test to verify, that would be useful.

rongpengzhang commented:
@mjwitte Yes, only (1) can be reported with the current settings in Win. (2) cannot.

Here are the test results:
![image](https://cloud.githubusercontent.com/assets/7785946/11882820/a2c5d144-a4c3-11e5-8ec2-ae198953b65f.png)
![image](https://cloud.githubusercontent.com/assets/7785946/11882872/07fdb57c-a4c4-11e5-8a7b-d4abeb33c78b.png)

FYI @Myoldmopar

rongpengzhang commented:
@mjwitte A bit more tests on the following approach:
1) In VS menu, Project -> Properties -> C/C++ -> Code Generation -> Floating Pint Model, change its setting to "Strict(/fp:strict)".
2) Use _controlfp to set the floating-point control word (see code below).

Here are the test results to handle "double x = 0.0; double y = 1 / x;":
(a) if I only implement 1), the FPE cannot be trapped
(b) if I then continue to implement 2), the FPE can be trapped
![image](https://cloud.githubusercontent.com/assets/7785946/11883215/417006dc-a4c6-11e5-85fd-9ca5b00feaa0.png)

mjwitte commented:
@rongpengzhang Thanks for testing.  Beware that the cmake scripts can override settings that you configure directly in VS - it's possible that the fp:strict is not actually being kept.  It's more reliable to add these to the cmake configuration by adding it to the CMAKE_CXX_FLAGS_DEBUG field.
